Dashboard migration: Add missing metrics registration

dprokop opened the pull request:
### Refactor dashboard migration metrics and logging with upstream implementation

#### Problem

Dashboard migration metrics registration was missing from the dashboard API service. Additionally, metrics and logging were hardcoded in individual conversion functions, leading to code duplication and inconsistent monitoring across different conversion paths.

#### Solution

- Register migration metrics in dashboard API service to enable performance tracking
- Move metrics reporting from individual conversion functions to upstream wrappers  
- Add schema migration metrics in migrate.go for all schema version changes
- Add conversion metrics wrapper in conversion.go for all API version changes
- Add new Prometheus metrics distinguishing schema migrations vs API conversions
- Implement withConversionMetrics wrapper pattern for automatic instrumentation

#### How to test

Run the migration tests to verify metrics are properly recorded:
```bash
go test -v ./apps/dashboard/pkg/migration
go test -v ./apps/dashboard/pkg/migration/conversion
```

dprokop committed:
Add dashboard migration metrics registration

• Register migration metrics in dashboard API service
• Enable tracking of dashboard conversion performance

stephaniehingtgen commented:
I think we'll also need to wire it up here: https://github.com/grafana/grafana/blob/1c587a983f855f1fbd478aeb7f476ffc4bc5a602/apps/dashboard/pkg/migration/migrate.go#L66 to increment counts, etc

dprokop committed:
Refactor dashboard migration metrics and logging to be upstream

- Move metrics reporting from individual conversion functions to upstream wrappers
- Add schema migration metrics in migrate.go for all schema version changes
- Add conversion metrics wrapper in conversion.go for all API version changes
- Create comprehensive test suite for both metrics and logging functionality
- Add new Prometheus metrics for schema migrations vs API conversions
- Implement withConversionMetrics wrapper pattern for automatic instrumentation

dprokop committed:
Merge remote-tracking branch 'origin/main' into dashboard-conversion-metrics

dprokop committed:
Merge remote-tracking branch 'origin/main' into dashboard-conversion-metrics

ivanortegaalba commented on apps/dashboard/pkg/migration/conversion/conversion.go:
```diff
@@ -9,76 +12,185 @@ import (
 var logger = logging.DefaultLogger.With("logger", "dashboard.conversion")
 
+// withConversionMetrics wraps a conversion function with metrics and logging for the overall conversion process
+func withConversionMetrics(sourceVersionAPI, targetVersionAPI string, conversionFunc func(a, b interface{}, scope conversion.Scope) error) func(a, b interface{}, scope conversion.Scope) error {
```
Can we rename `a` and `b` to understand they are dashboards? wdyt?

ivanortegaalba replied:
What about extracting this function into a different file? So we keep this one as registry :)

ivanortegaalba commented on apps/dashboard/pkg/migration/conversion/conversion.go:
```diff
@@ -9,76 +12,185 @@ import (
+		case *dashv2alpha1.Dashboard:
+			targetSchemaVersion = "v2alpha1"
+		case *dashv2beta1.Dashboard:
+			targetSchemaVersion = "v2beta1"
```
WDYT about not defining `sourceSchemaVersion` and `targetSchemaVersion` for conversions different from v0 -> v1? So we avoid errors using schemaVersions instead of apiVersions for monitoring conversions.

ivanortegaalba commented on apps/dashboard/pkg/migration/conversion/conversion.go:
```diff
@@ -9,76 +12,185 @@ import (
-	if err := s.AddConversionFunc((*dashv2beta1.Dashboard)(nil), (*dashv2alpha1.Dashboard)(nil), func(a, b interface{}, scope conversion.Scope) error {
-		return Convert_V2beta1_to_V2alpha1(a.(*dashv2beta1.Dashboard), b.(*dashv2alpha1.Dashboard), scope)
-	}); err != nil {
+	if err := s.AddConversionFunc((*dashv2beta1.Dashboard)(nil), (*dashv2alpha1.Dashboard)(nil),
```
I love this pattern 💯

ivanortegaalba reviewed (approved):
Codewise looks great to me! 

I’m just curious, do we really need double reporting metrics 🤔?  We're reporting schemaVersionMetric together with the conversion metric.

Schema migrations are simply part of the conversion from v0 to v1. They happen whenever a dashboard from v0 is requested from a different endpoint. 

Maybe we could treat schemaMigrations as additional metadata for the conversion metric from v0 to v1? What’s the reason for needing the schemaVersion metric specifically?

dprokop committed:
Merge remote-tracking branch 'origin/main' into dashboard-conversion-metrics

dprokop committed:
WIP: Optimize dashboard migration failure tracking to reduce costs

• Remove migration-specific metrics and eliminate redundant reporting
• Add function name tracking to MigrationError for better debugging
• Exclude schema versions from v2+ dashboard logs (redundant with API versions)
• Change schema_minimum_version_error from error to warning level

ivanortegaalba committed:
Implement ConversionError with API version fields and improve code organization

- Create ConversionError type following MigrationError pattern with currentAPIVersion and targetAPIVersion fields
- Update error handling to use exact conversion function names instead of generated ones
- Use version-only format (v2alpha1) instead of full API version (dashboard.grafana.app/v2alpha1)
- Match MigrationError argument order: msg, currentVersion, targetVersion, functionName
- Extract withConversionMetrics and related functions to metrics.go for better code organization
- Clean up conversion.go to focus only on RegisterConversions
- Add comprehensive tests for new error handling and function extraction
- Ensure metrics and logs show exact conversion function names like ConvertDashboard_V2alpha1_to_V2beta1

dprokop committed:
Add docs

harisrozajac committed:
lint

harisrozajac committed:
Merge branch 'main' into dashboard-conversion-metrics
